src/utils/validation/expressionValidation.ts

```diff
@@ -0,0 +1,189 @@
+import { evalExpr } from 'src/features/expressions';
+import { ExprVal } from 'src/features/expressions/types';
+import { getBaseDataModelBindings } from 'src/utils/databindings';
+import { buildValidationObject } from 'src/utils/validation/validationHelpers';
+import type { ExprConfig } from 'src/features/expressions/types';
+import type { IFormData } from 'src/features/formData';
+import type { LayoutNode } from 'src/utils/layout/LayoutNode';
+import type {
+  IExpressionValidation,
+  IExpressionValidationConfig,
+  IExpressionValidationRefResolved,
+  IExpressionValidationRefUnresolved,
+  IExpressionValidations,
+  IValidationContext,
+  IValidationObject,
+} from 'src/utils/validation/types';
+
+/**
+ * Resolves a reusable expression validation definition.
+ */
+function resolveExpressionValidationDefinition(
+  name: string,
+  definition: IExpressionValidationRefUnresolved,
+  config: { [name: string]: IExpressionValidationRefResolved },
+): IExpressionValidationRefResolved | null {
+  let resolvedDefinition = definition;
+
+  if ('ref' in definition) {
+    const reference = config[definition.ref];
+    if (!reference) {
+      window.logWarn(
+        `Custom validation:\nTried to reference ${definition.ref} in ${name} but it does not exist. Make sure it is defined before it is referenced.`,
+      );
+      return null;
+    }
+
+    const { ref: _, ...definitionWithoutRef } = definition;
+    resolvedDefinition = { ...reference, ...definitionWithoutRef };
+  }
+
+  if (!('message' in resolvedDefinition)) {
+    window.logWarn(`Custom validation:\nDefinition for ${name} is missing a message.`);
+    return null;
+  }
+
+  if (!('condition' in resolvedDefinition)) {
+    window.logWarn(`Custom validation:\nDefinition for ${name} is missing a condition.`);
+    return null;
+  }
+
+  return resolvedDefinition as IExpressionValidationRefResolved;
+}
+
+/**
+ * Resolves a single expression validation definition.
+ */
+function resolveExpressionValidation(
```

Comment:
This function, and the one above it, seems to duplicate some things. Looks a bit strange, am I missing a reason the duplication is needed? 🤔

Reply:
Not entirely sure what you mean, but then again I have not looked at this function for weeks. I will take a look

Reply:
Both of them have things like:
```
 if (!('message' in ...)) {
    window.logWarn(`Custom validation:\nValidation for ${field} is missing a message.`);
    return null;
  }
```

Reply:
They deal with the`validations` and `definitions` object (from the config) respectively, as they are treated slightly differently. But I have them both require a message and condition to be set at a minimum